apm-lambda-extension/e2e-testing/e2e_test.go

```diff
@@ -0,0 +1,289 @@
+var langPtr = flag.String("lang", "node", "the language of the Lambda test function : Java, Node, or Python")
+var timerPtr = flag.Int("timer", 20, "the timeout of the test lambda function")
+var javaAgentVerPtr = flag.String("java-agent-ver", "1.28.4", "the version of the java APM agent")
+
+func TestEndToEnd(t *testing.T) {
+
+	// Check the only mandatory environment variable
+	if err := godotenv.Load(".e2e_test_config"); err != nil {
+		log.Println("No additional .e2e_test_config file found")
+	}
+	if getEnvVarValueOrSetDefault("RUN_E2E_TESTS", "false") != "true" {
+		t.Skip("Skipping E2E tests. Please set the env. variable RUN_E2E_TESTS=true if you want to run them.")
+	}
+
+	supportedLanguages := []string{"node", "python", "java"}
+	if !isStringInSlice(*langPtr, supportedLanguages) {
+		processError(errors.New("unsupported language"))
+	}
+
+	samPath := "sam-" + *langPtr
+	samServiceName := "sam-testing-" + *langPtr
+
+	// Build and download required binaries (extension and Java agent)
+	buildExtensionBinaries()
+
+	// Java agent processing
+	if *langPtr == "java" {
+		if !folderExists(filepath.Join(samPath, "agent")) {
+			log.Println("Java agent not found ! Collecting archive from Github...")
+			retrieveJavaAgent(samPath, *javaAgentVerPtr)
+		}
+		changeJavaAgentPermissions(samPath)
+	}
+
+	// Initialize Mock APM Server
+	mockAPMServerLog := ""
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.RequestURI == "/intake/v2/events" {
+			bytesRes, _ := getDecompressedBytesFromRequest(r)
+			mockAPMServerLog += string(bytesRes)
+		}
+	}))
+	defer ts.Close()
+
+	resultsChan := make(chan string, 1)
+
+	uuid := runTestWithDedicatedTimer(samPath, samServiceName, ts.URL, *rebuildPtr, *timerPtr, resultsChan)
+	if uuid == "" {
+		t.Fail()
+	}
+	log.Printf("Querying the mock server for transaction bound to %s...", samServiceName)
+	assert.True(t, strings.Contains(mockAPMServerLog, uuid))
+}
+
+func runTestWithDedicatedTimer(path string, serviceName string, serverURL string, buildFlag bool, timeout int, resultsChan chan string) string {
+	timerNode := time.NewTimer(time.Duration(timeout) * time.Second * 2)
+	go runTest(path, serviceName, serverURL, buildFlag, timeout, resultsChan)
+	select {
+	case uuid := <-resultsChan:
+		return uuid
+	case <-timerNode.C:
+		break
+	}
+	return ""
+}
+
+func buildExtensionBinaries() {
+	runCommandInDir("make", []string{}, "..", getEnvVarValueOrSetDefault("DEBUG_OUTPUT", "false") == "true")
+}
+
+func runTest(path string, serviceName string, serverURL string, buildFlag bool, timeout int, resultsChan chan string) {
+	log.Printf("Starting to test %s", serviceName)
+
+	if !folderExists(filepath.Join(path, ".aws-sam")) || buildFlag {
+		log.Printf("Building the Lambda function %s", serviceName)
+		runCommandInDir("sam", []string{"build"}, path, getEnvVarValueOrSetDefault("DEBUG_OUTPUT", "false") == "true")
+	}
```

Comment:
If I'm reading this correctly, if there is no lambda function already built and the `buildFlag` is `false`, we'll skip building the lambda function and still end up trying to run it below. We should probably exit here if there is no lambda function built and the `buildFlag` is `false`.

Reply:
Actually, we enter the if statement if no Lambda image is built OR if buildFlag is true. If there is no Lambda image, we will always build one, no matter the value of `buildFlag`

Reply:
of course 🤦🏻‍♀️
sorry, I read the logic wrong